### PR TITLE
Add ReserveSpace in dygraph batch_norm.

### PR DESCRIPTION
**Pull request by huangxu96**
### PR types

Function optimization
### PR changes

APIs
### Describe

This PR added ReserveSpace flag for BatchNorm in dygraph model for supporting fp16 training.

### CODE REVIEW

File: python/paddle/fluid/dygraph/nn.py [hunk 1]
```diff
@@ -1308,6 +1309,12 @@ def __init__(self,
             dtype=self._dtype)
         self._variance.stop_gradient = True
 
+        self._has_reserve_space = False
+        if data_layout == 'NHWC':
+            flag = os.environ.get('FLAGS_cudnn_batchnorm_spatial_persistent')
```
Comment (zhiqiu) [inline python/paddle/fluid/dygraph/nn.py:1314]: Why not use paddle.get_flags instead of os.environ.get

Comment (huangxu96) [reply python/paddle/fluid/dygraph/nn.py:1314]: OK，I will fix it in next PR

**Review by wzzju — APPROVED** [review]
LGTM.
